src/action_exec.ml

```diff
@@ -108,7 +108,14 @@ let rec exec t ~ectx ~dir ~env ~stdout_to ~stderr_to =
     Path.rm_rf path;
     Fiber.return ()
   | Mkdir path ->
-    Path.mkdir_p path;
+    begin
+      if Path.is_in_build_dir path then
+        Path.mkdir_p path
+      else if Path.exists path then
+        ()
+      else
+        die "action attempted to create %a outside the build dir" Path.pp path
```

Comment:
It's a bit unfortunate that we can't give the user a proper location here, but passing the location along would be quite painful